zerver/tests/test_events.py

```diff
@@ -1115,20 +1115,20 @@ def test_register_events_email_address_visibility(self) -> None:
 
         events = self.verify_action(lambda: self.register("test1@zulip.com", "test1"), num_events=5)
         self.assert_length(events, 5)
-        check_realm_user_add("events[0]", events[0])
+        check_realm_user_add("events[1]", events[1])
```

Comment:
These test changes are somewhat weird. Actual order of events is not changed. But the problem occurs in tests when all events are not sent using `transaction.on_commit`. In actual code if `on_commit` is called when outside a transaction, the code is executed immediately.
But in tests, if for a particular case some events are inside `on_commit` (due to it being used inside a transaction by some other code) and others are not, the `on_commit` callbacks are executed at last as these tests are wrapped by `captureOnCommitCallbacks(execute=True)`.

Reply:
Ofcourse, this is an intermediate state and will be fixed once we move other events to be called inside `on_commit`.

Reply:
I think this will happen in real life too -- we are actually changing the order of the events. And I think that could cause issues for lcients.

I think probably what we need to do is first move the events that were previously at the end to have `transaction.on_commit` wrappers. For example, make the "send a message to the signups stream" thing use `transaction.on_commit` first, to avoid the reordering that is happening here.

Reply:
Yeah, structurally, a correct Zulip codebase using transactions for every actions function would always do `send_event` calls inside `transaction.on_commit`. That is usually equivalent to doing the `send_event` just after the transaction, but that's more an implementation that happens to be equivalent to the correct implementation.

Reply:
I guess what I was trying to say is I'm still really confused about why the event ordering is changing for these tests. It sounds like the theory is something about the interleaving of calls to `send_event` inside `transaction.on_commit` with `send_event` calls outside it, but I would have thought all the `send_event` calls would be inside `transaction.on_commit` already.

I don't like just changing the tests to expert events to occur out of order from how they'll appear in reality, because that's making the test verify incorrect behavior, and in at least some cases I'd expect client code to break if the sequence of events being generated by this flow were delivered in a random order. So I'd really like us to figure out a way to make the tests verify that the order is as it appears in the code, and that order matches reality.